Add controller StopFunc support

raybejjani opened the pull request:
**Summary of changes**:
***Background***
I need to do some resource cleanup when an endpoint is deleted. The state is syncronised via a controller. The current controller design doesn't allow us to use a defer for this, since the user provided DoFunc is called repeatedly (vs the whole controller being a goroutine we write).

***Changes***
- When DoFunc is not set, it will be set to an appropriate stub function. This avoids some extra nil checks in the loop code of the controller
- StopFunc is called on a controller exit.
- Some small contoller refactoring of stats collection to make StopFunc work consistently

***Things to consider***
- Should StopFunc be a defer? We would then wrap the whole shutdown section into a deferred function. This would allow it to catch panics and handle those, but that may not be too useful.
- Does a StopFunc even make sense? A controller is rather long lived, and in the use case I have in mind (endpoint.LeaveLocked) we can install a delete controller instead, so maybe it isn't needed.

In support of https://github.com/cilium/cilium/issues/2183


**How to test (optional)**:
none. I did add a unit test, however.

raybejjani committed:
controller: Simplify DoFunc init handling

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
controller: Add StopFunc on controller exit

When a controller exits it might need to cleanup or indicate that it
exited. StopFunc is called as the controller shuts down.

Signed-off-by: Ray Bejjani <ray@covalent.io>

tgraf commented on pkg/controller/controller.go:
```diff
@@ -34,6 +35,14 @@ type ControllerParams struct {
+	// clean-up tasks for the controller (e.g. deallocate/release resources)
+	// It is guaranteed that DoFunc is called at least once before StopFunc is
+	// called.
+	// Note: Since this occurs on controller exit, error counts and tracking my
```
may

raybejjani replied:
fixed

tgraf reviewed (changes requested):
Looks great, just a small typo.

raybejjani committed:
controller: Do not reuse controller struct & stats

When updating a controller we reusued the existing struct. The docs
state that we reset the statistics but this usage does not. It also
allows for a race where the running controller will lookup c.params but
this has now changed as part of the update.

Signed-off-by: Ray Bejjani <ray@covalent.io>
